### ISSUE TITLE
Lacking built-in variables

### ISSUE THREAD

**Issue by jilleJr** [iver-wharf/wharf-cmd#55]
We need to be able to provide built-in variables (https://iver-wharf.github.io/#/usage-wharfyml/variables/built-in-variables) in both `.wharf-ci.yml` parsing as a variable substitution source as well as in the step types as some have built-in var depending default values, such as `docker.group` step type field (https://iver-wharf.github.io/#/usage-wharfyml/step-types/docker) that should default to `REPO_GROUP`.

Some variables are more environment focused, such as:

- `REG_URL`
- `DEFAULT_DOMAIN`
- `CHART_REPO`

While others are build-specific, such as:

- `BUILD_REF`
- `GIT_COMMIT`
- `REPO_GROUP`

Need to plan out how to provide these built-in vars in a good way that works both locally and from the Kubernetes perspective.

One idea is to have `.env` files of values, where we could let wharf-cmd look at multiple and merge the values.

Such as:

```env
# ~/.config/iver-wharf/wharf-cmd/built-in-vars.env
CHART_REPO=https://harbor.local/library
```

```env
# /tmp/iver-wharf/wharf-cmd/build/123/built-in-vars.env
BUILD_REF=123
GIT_COMMIT=9a634cb05824565110a119bd60ef6e1c8243bb40
```

Suggest then that wharf-cmd looks at:

- `~/.config/iver-wharf/wharf-cmd/built-in-vars.env`
- `/etc/iver-wharf/wharf-cmd/built-in-vars.env`
- All `.wharf-cmd-built-in-vars.env` files in parent directories, recursively

And uses those to auto-generate `/tmp/iver/wharf-cmd/build/123/built-in-vars.env`, which would be what it's actually using in the build.

With this, we could provide a flag such as `--var-env-file=file.env` that adds another place for Wharf to look for these files, and let the variables be populated by Kubernetes secrets and whatnot.

Kubernetes secrets and configmaps are mounted as each file as a separate secret value, so perhaps add support for that instead of forcing `.env` file would be a good idea.

Also, YAML is probably a better format than `.env` anyway, as it preserves the types.

---

If it wasen't clear from the rambling issue description, this definetly needs some dedicated planning time before coding starts.

**Comment by jilleJr** [comment]
Plan:

- Use the newly created `pkg/gitstat` to populate all git-related built-in vars
- Try get `REPO_NAME` and `REPO_GROUP` by parsing the git remote
- Use YAML files for built-in vars, allowing base values and overrides. Suggest to have wharf look at `~/.config/iver/wharf/built-in-vars.yml`, `/etc/iver/wharf/built-in-vars.yml`, and all `.wharf-vars.yml` files recursively up from the `.wharf-ci.yml`'s path.
- Validate on `wharf-cmd run` for missing built-in vars
- Track latest `BUILD_REF` in `~/.local/share/iver/wharf/last-build.yml` and use that as a baseline.

This allows wharf-cmd-provisioner to populate wharf-cmd-worker with built-in vars by providing a `.wharf-vars.yml`/`built-in-vars.yml` file.

These var sources needs to be appropriately merged to allow values to override each other. Ex `.wharf-vars.yml` should override with higher prio the closer to the `.wharf-ci.yml` file it is.

**Comment by jilleJr** [comment]
So in total, we would have:

## `.wharf-ci.yml`

Build definition.

- only at the repo root.

Sample:

```yaml
# .wharf-ci.yml
environments:
  myEnv: {}

inputs: []

myStage:
  myStep:
    container:
      image: ubuntu
      cmds: [echo lorem]
```

## `.wharf-vars.yml`

Variable definitions.

- at any path in or above the same dir as `.wharf-ci.yml`.

Sample:

```yaml
# .wharf-vars.yml
vars:
  CHART_REPO: foo
```

## `.wharf-conf.yml`

Variable definitions and wharf-cmd specific configs

- at any path above the same dir as `.wharf-ci.yml`, but not the same dir
- at `~/.config/iver-wharf/wharf-cmd/wharf-conf.yml` (GNU/Linux)
- at `/etc/iver-wharf/wharf-cmd/wharf-conf.yml` (GNU/Linux)
- or at `%APPDATA%/iver-wharf/wharf-cmd/wharf-conf.yml` (Windows)

Sample:

```yaml
# .wharf-conf.yml
runner: kubernetes
kubernetes:
  kubeconfig: ../.kube/config
vars:
  CHART_REPO: foo
```

---

This needs an RFC...

**Comment by jilleJr** [comment]
Maybe better to just merge `.wharf-conf.yml` and `.wharf-vars.yml` anyway, but then in `wharf-cmd run` validate that you are only setting `vars` in the `.wharf-conf.yml` file inside the same dir as the `.wharf-ci.yml`.

One less file, and the error messages could be made useful, as well as opening up to allow _some_ configs to be declared inside the repo. Can't come up with any obvious, but it does make it somewhat future-proofer

**Comment by Alexamakans** [comment]
Good stuff here 👍🏻

---
1.
> However from a security perspective, allow a Git repo to define such things could be a big nono.

I kind of think that it is the developer's responsibility to not upload sensitive data, and maybe we should still support it at the same level. But I think that that might be my bias speaking because that's how it's generally been done in projects I've worked in before.

If disallowing it completely, then warning/erroring if it is detected next to the `wharf-ci.yml` file or in the repo with something kinda like:
```
Found `.wharf-conf.yml` file in repository (or next to the wharf-ci file idk), but it will not be loaded.
This file may contain secret data and should not be there. Please refer to [some link] for more information.
```
maybe?

---
2.
> [Merge them, and have] `wharf-cmd run` validate that you are only setting vars in the `.wharf-conf.yml` file inside the same dir as the `.wharf-ci.yml`.

I think it would be fairly confusing to have the same type of file have different behaviors depending on its location relative to another file. I vote for separating them entirely. It would be less error-prone and cut down on development time as well I think.

I.e. only vars in `.wharf-vars.yml`, and only config stuff for `.wharf-conf.yml`.
Unless having vars in `.wharf-conf.yml` makes things a lot easier that is, but ideally I would have them be separate.

**Comment by jilleJr** [comment]
@Alexamakans The security aspect are relevant for users that don't know all about wharf, which will be a lot of users.

Imagine a simple README on a foreign repo explaining a build where you clone a repo and then just do `wharf run`.

If that repo can control what kubeconfig context to use from your personal kubeconfig, then that has the possibility with a slight hack to expose the users kubeconfig.

But idk, in that case it doesn't stop the author much from having a `hackme.sh` file that does it much more cleanly instead. Would be similarly easy to fool the uninformed. 

However it's not enough argument to not add yet another vector if we can avoid it.

Another concern is the bloat of files. Preferably the .wharf-ci.yml file would be more extensible in its format to allow these settings right inside the file, and then only have .wharf-ci.yml files that depend on each other.

**Comment by Alexamakans** [comment]
> If that repo can control what kubeconfig context to use from your personal kubeconfig, then that has the possibility with a slight hack to expose the users kubeconfig.

Good point.

> [Merge them, and have] wharf-cmd run validate that you are only setting vars in the .wharf-conf.yml file inside the same dir as the .wharf-ci.yml.

I realized now that `validate` here perhaps was referring to the warning/erroring I mentioned in my previous comment.

> Preferably the .wharf-ci.yml file would be more extensible in its format to allow these settings right inside the file, and then only have .wharf-ci.yml files that depend on each other.

This would be preferable indeed. It also makes me feel a bit more okay, for some reason, with having the ability to set config values when outside the main directory.

Probable that implementation of this would be better (dare I say, easier, even? Maybe... 🤔) as well, in that we have a good visitor-pattern with line errors in place now.

**Comment by jilleJr** [comment]
New plan: Only `wharf-conf` file in config dir. Not in dir tree.

```
~/.config/iver-wharf/wharf-cmd/wharf-conf.yaml
~/.config/iver-wharf/wharf-cmd/wharf-vars.yaml
/etc/iver-wharf/wharf-cmd/wharf-conf.yaml
/etc/iver-wharf/wharf-cmd/wharf-vars.yaml

./.wharf-vars.yaml
../.wharf-vars.yaml
../../.wharf-vars.yaml
../../../.wharf-vars.yaml

./.wharf-ci.yml
```

**Comment by Alexamakans** [comment]
```yaml
# .wharf-conf.yml
runner: my-kubernetes
runners:
  my-kubernetes:
    type: kubernetes
    kubeconfig: ../.kube/config
vars:
  CHART_REPO: foo
```

**Comment by jilleJr** [comment]
```yaml
runner:
  current: kubernetes
  kubernetes:
    kubeconfig: ../.kube/config
```

**Comment by jilleJr** [comment]
Tried to summarize some meeting notes here: https://iver-wharf.github.io/wharf-notes/2022-03-11-wharf-cmd-built-in-vars

The config part with presets is fine if we postpone. It's a nice idea, though I would like to delve deeper into it before we start implementing.

This issue is only regarding the built-in variables, and it seems like the solution forward is to simply add a `.wharf-vars.yml` file anyway. We can extend with a `~/.config/iver-wharf/wharf-cmd/wharf-conf.yml` file later.

Still probably needs a new name for `.wharf-vars.yml` though.

**Comment by jilleJr** [comment]
Possible other names for `.wharf-vars.yml`, since "Wharf vars" is so awkward to pronounce:

1. `.wharf-env.yml`
2. `.wharf-builtins.yml`
3. `.wharf-user.yml`
4. `.wharf-base.yml`
5. `.wharf-variables.yml`

`.wharf-variables.yml` collides the least with other terms inside Wharf as well as semantically with other concepts.

`.wharf-builtins.yml` is the most explicit when referring to Wharf's built-in variables, but they are not "built in" any more, so a rename might be in order, and then the variable file should have a broader name.

Or we keep `.wharf-vars.yml` but emphasize that it's pronounced "Wharf variables YML"

**Comment by Alexamakans** [comment]
I vote for `.wharf-variables.yml` or `.wharf-vars.yml`, clarity of the name should precede ease of pronunciation imo.
